tests/python/relax/test_dataflow_inplace.py

```diff
@@ -0,0 +1,464 @@
+        for i0, i1 in T.grid(T.int64(2), T.int64(3)):
+            with T.block("compute"):
+                v_i0, v_i1 = T.axis.remap("SS", [i0, i1])
+                T.reads(A[v_i0, v_i1])
+                T.writes(compute[v_i0, v_i1])
+                compute[v_i0, v_i1] = T.sigmoid(A[v_i0, v_i1])
+        for ax0, ax1 in T.grid(T.int64(2), T.int64(3)):
+            with T.block("T_multiply"):
+                v_ax0, v_ax1 = T.axis.remap("SS", [ax0, ax1])
+                T.reads(A[v_ax0, v_ax1], compute[v_ax0, v_ax1])
+                T.writes(A[v_ax0, v_ax1])
+                A[v_ax0, v_ax1] = A[v_ax0, v_ax1] * compute[v_ax0, v_ax1]
+
+    silu_call = TestModule["main"].body.blocks[0].bindings[1].value
+    new_silu, new_mod = dataflow_single_inplace_call(TestModule, silu_call, [0])
+
+    tvm.ir.assert_structural_equal(new_mod["silu_inplace"], expected_silu)
+    assert new_silu.op.name == "relax.call_tir_inplace"
+    assert new_silu.args[0].name_hint == "silu_inplace"
+    for i, arg in enumerate(new_silu.args[1].fields):
+        arg == silu_call.args[i]
+    new_silu.attrs.inplace_indices == [0]
+
+
+def test_insert_inplace_calls():
+    @I.ir_module
+    class EndToEndTest:
+        @R.function
+        def main(
+            x: R.Tensor((2, 3), dtype="float32"), y: R.Tensor((1, 3), dtype="float32")
+        ) -> R.Tensor((2, 3), dtype="float32"):
+            with R.dataflow():
+                z = R.add(x, y)  # broadcast happens here
+                # Cannot be done in-place because x is an argument.
+                a = R.add(z, y)  # this one can be done in-place
+                q = R.multiply(a, y)  # broadcast again, a is eligible
+                r = R.subtract(y, y)  # cannot be done in-place because y is an argument
+                s = R.subtract(r, r)  # No broadcast. Can be done in-place
+                m = R.multiply(q, s)  # should give us all zeros
+                R.output(m)
+            return m
+
+    transform_pass = DataflowUseInplaceCalls()
+    new_mod = transform_pass(EndToEndTest)
+
+    # check that all operations are done in-place
+    assert new_mod["add_inplace"]
+    assert new_mod["subtract_inplace"]
+    assert new_mod["multiply_inplace"]
+    expected_ops = ["add_inplace", "multiply_inplace", "subtract_inplace", "multiply_inplace"]
+    inplace_binding_idxs = [1, 2, 4, 5]
+    for i, idx in enumerate(inplace_binding_idxs):
+        binding = new_mod["main"].body.blocks[0].bindings[idx]
+        assert binding.value.op.name == "relax.call_tir_inplace"
+        assert binding.value.args[0].name_hint == expected_ops[i]
+
+    x = tvm.nd.array(np.random.rand(2, 3).astype("float32"))
+    y = tvm.nd.array(np.random.rand(1, 3).astype("float32"))
+    expected = np.zeros((2, 3), dtype="float32")
+
+    target = tvm.target.Target("llvm")
+    ex = relax.build(new_mod, target)
+    vm = relax.VirtualMachine(ex, tvm.cpu())
+    res = vm["main"](x, y)
+    assert (expected == res.numpy()).all()
+
```

Comment:
Do we want a `pytest.xfail` case for internal functions, to indicate that they are not yet supported?  I'm thinking of cases like the following, where the current local analysis would conclude that `Module.func` cannot do in-place operations, but non-local analysis could perform `Module.func` in-place for some cases, depending on the liveness analysis in the calling scope.

```python
@I.ir_module
class Module:
    @R.function
    def main(x: R.Tensor((), "int32")) -> R.Tensor((), "int32"):
        cls = Module
        with R.dataflow():
            # Cannot be in-place, uses a function argument
            y = cls.subroutine(x)
            
            # Can be in-place, uses a local binding
            z = cls.subroutine(y)
            R.output(z)
        return z
    
    @R.function(private=True)
    def subroutine(x: R.Tensor((), "int32"), y: R.Tensor((), "int32")) -> R.Tensor((), "int32"):
        with R.dataflow():
            z = R.add(x,R.const(1, "int32"))
            R.output(z)
        return z
```

Reply:
I'm always in favor of testing error conditions, so I'll add the test case too.

Reply:
Sounds good!  Though, I'd probably see this one as a not-yet-supported test case (`pytest.mark.xfail`), rather than a test of an error case (`with pytest.raises(...)`).  Rather than verifying that a specific error is raised, it would document what behavior isn't covered by the current implementation.

Reply:
On second thought, I'm not sure we should add an error-condition test case for something we don't even claim to support (the description of the pass does not reference trying to do function calls in place).

Reply:
I could go either way on it.  The high-level name `DataflowUseInplaceCalls` doesn't specify that the `InplaceCalls` only refer to native relax operators, and excludes potentially-in-place calls to other relax functions.  The docstring does specify operators, so you're right that we don't claim support for it there.

I'd see a `@pytest.mark.xfail(reason="Not currently supported")` as a way to explicitly disclaim responsibility for that use case.  It would indicate what that support would look like, and that it could be supported at some point, but that it isn't currently supported.

Reply:
I can appreciate your point of view here, but I would point out that while tests are a form of documentation, documentation (such as doc comments) is also a form of documentation :slightly_smiling_face: 

If there is a strong need to reinforce the point that we are only looking for a small number of operators, we can do this, but I don't think there is right now, especially since handling Relax functions in-place is not something we have even planned on doing. (Possible in principle, but it's not on the horizon.)

Reply:
Good point, and mostly just nitpicks on my side.  While I think documenting potential future expansions is useful, I don't think it's worth delaying this PR.  (Approved!)